New target future sequana

lrusinowicz opened the pull request:
### Description
@MarceloSalazar

This commit adds support for Future Electronics Sequana target (FUTURE_SEQUANA) including basic support for Cypress PSoC 6 microcontrollers.
Tests were run on a Sequana board and are all passing for all 3 compilers (GCC_ARM, ARM, and IAR).

[tests_sequana_arm_20181020.log](https://github.com/ARMmbed/mbed-os/files/2501448/tests_sequana_arm_20181020.log)
[tests_sequana_gcc_20181002.log](https://github.com/ARMmbed/mbed-os/files/2501449/tests_sequana_gcc_20181002.log)
[tests_sequana_iar_20181020.log](https://github.com/ARMmbed/mbed-os/files/2501450/tests_sequana_iar_20181020.log)

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

Nodraak committed:
Initial support for PSoC 6.

lrusinowicz committed:
Initial version with support for PSoC 6 CY8C63XX and both cores using updated PDL 3.0.1 beta.

lrusinowicz committed:
Updated TARGET_Cypress sub-tree for FUTURE_SEQUANA target:

1. Complete set of HAL drivers for all peripherals of CY8C63xx PSoC chips.
2. Cypress PDL library updated to official 3.0.1 version.
3. Tree structure reorganized and cleaned up:
 + TARGET_Cypress
 +--+ TARGET_PSOC6+                  -> code & libs applicable to all PSoC 6 based devices
    +--+ TARGET_CY86XX               -> code & libs applicable to PSoC 63 based devices
    |  +--- TARGET_MCU_PSOC6_M0      -> code & libs applicable to PSoC6 Corted M0+ core
    |  +--- TARGET_MCU_PSOC6_M4      -> code & libs applicable to PSoC6 Corted M0F core
    |
    +--+ TARGET_FUTURE_SEQUANA       -> code applicable to Sequana board, both cores
       +--- TARGET_FUTURE_SEQUANA_M0 -> code applicable only to M0+ core on Sequana board

lrusinowicz committed:
Added BLE support based on CORDIO stack.

lrusinowicz committed:
Added required changes outside of TARGET_Cypress tree:

1. In drivers/Timer.cpp make sure that hardware timer is initialized outside of critical section.
   This is because on PSoC 6 hardware resources are shared between both cores
   and we have to make sure that the other core is not already using a particular resource.
   This mechanism is based on interprocessor communication taht cannot be handled iside of
   critical section.
2. Added support for post-binary hook function for PSoC 6 targets, so the hex image for M0+ CPU core
   can be merged with M4 core image for the final image.
3. Added possibility to use hook function from exportes, so the M0+ hex image could be included
   in the generated project.
4. Included hex images in the build dependency list, so the update of image is catched by the
   build process.

theotherjimmy commented on tools/export/__init__.py:
```diff
@@ -129,6 +131,62 @@ def mcu_ide_matrix(verbose_html=False):
+    # "classname.functionname"
+    temp = hook_data["export_function"].split(".")
+    if len(temp) != 2:
+        raise HookError(
```
Use of HookError before its definition.

theotherjimmy commented on tools/export/__init__.py:
```diff
@@ -129,6 +131,62 @@ def mcu_ide_matrix(verbose_html=False):
+
+    if class_name not in mdata or \
+       not inspect.isclass(mdata[class_name]):
+        print (mdata)
```
Debug print left in.

theotherjimmy commented on tools/export/__init__.py:
```diff
@@ -129,6 +131,62 @@ def mcu_ide_matrix(verbose_html=False):
+       not toolchain_labels.intersection(toolchain_restrictions):
+        return None
+    # Finally, get the requested function
+    print("Found function '%s' in class '%s'."% (function_name, class_name))
```
Dedub print left in.

theotherjimmy commented on tools/export/__init__.py:
```diff
@@ -129,6 +131,62 @@ def mcu_ide_matrix(verbose_html=False):
+    Positional Arguments:
+    target - the target object for inspection
+    toolchain - the toolchain object for inspection
+    """
```
This function looks like it's copied from https://github.com/ARMmbed/mbed-os/blob/master/tools/targets/__init__.py#L330-L380. Please don't copy code.

theotherjimmy commented on tools/export/__init__.py:
```diff
@@ -158,6 +216,9 @@ def generate_project_files(resources, export_path, target, name, toolchain, ide,
     exporter = exporter_cls(target, export_path, name, toolchain,
                             extra_symbols=macros, resources=resources)
+    # Locate target-specyfic exporter if specified.
+    exporter.TARGET_EXPORTER = get_target_exporter(target, toolchain)
```
Setting a class-constant is generally not a good idea. This should be passed into the constructor, or at least an object member.

theotherjimmy commented on tools/export/exporters.py:
```diff
@@ -53,7 +53,7 @@ class Exporter(object):
     TOOLCHAIN = None
     CLEAN_FILES = ("GettingStarted.html",)
-
+    TARGET_EXPORTER = None
```
This changes with each _object_, so it must be initialized during the constructor.

theotherjimmy commented on tools/export/makefile/__init__.py:
```diff
@@ -145,6 +146,12 @@ def generate(self):
 
+        # If there is a target-specific exporter function, call it now.
+        if self.TARGET_EXPORTER:
+            print ("Calling target exporter...")
```
Debug print left in.

theotherjimmy commented on tools/export/makefile/__init__.py:
```diff
@@ -154,6 +161,7 @@ def generate(self):
             ['makefile/%s.tmpl' % self.TEMPLATE]:
             try:
                 self.gen_file(templatefile, ctx, 'Makefile')
+                print("Generated Makefile from %s"%templatefile)
```
Debug print left in.

theotherjimmy commented on tools/export/makefile/__init__.py:
```diff
@@ -145,6 +146,12 @@ def generate(self):
+        # If there is a target-specific exporter function, call it now.
+        if self.TARGET_EXPORTER:
+            print ("Calling target exporter...")
+            self.TARGET_EXPORTER(self, ctx)
```
As this is the only exporter that uses this "feature" I'm not convinced that it's really needed. Could you explain why Makefiles are the only exporter that need to merge the Hex file? Further, https://github.com/ARMmbed/mbed-os/pull/8491/files#diff-a35f26ab826f25fd621856dc9d866598R94 and https://github.com/ARMmbed/mbed-os/pull/8491/files#diff-a35f26ab826f25fd621856dc9d866598R95 indicate that there will be overlapping regions in these merged hex files. This will cause the Makefile exporter to always export a non-functional Makefile for your target.

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+from ..config import ConfigException
+
+
+class HookError(Exception):
```
Please import this instead of defining your own. This makes it quite a bit more difficult to catch your `HookError`.

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+    for start, end in ihex.segments():
+        if start  >= 0x090000000:
+            continue
+        segment = ihex.tobinarray(start = start, end = end)
```
Generally and elsewhere in this file, we don't put ` ` around `=` in the argument positions.

theotherjimmy commented on tools/export/makefile/__init__.py:
```diff
@@ -54,7 +54,8 @@ class Makefile(Exporter):
         "LPCTargetCode.lpc_patch",
-        "LPC4088Code.binary_hook"
+        "LPC4088Code.binary_hook",
+        "PSOC6Code.complete"
```
Considering that this adds a checksum to the binary, does the device boot without this checksum? If not, please remove any exporters that can't/don't edit this checksum value, including make.

lrusinowicz replied:
Actually, this checksum is only used/verified by Cypress programming tools (PSoC Programmer, KitProg2). Our new, DAPLink based interface simply ignores it, so there is no longer strict requirement to have a checksum calculated at build time . Nevertheless, I've decided to retain these calculations to have a final image compatible with a broader range of development tools. 
To answer your question - exported projects do work, albeit with limitations similar to bootloader case. One would usually have to resort to some other means of uploading required M0 image (except for makefile project, which can build combined image using srec_cat tool as for the Nordic chips).
As our interface does not perform 'whole chip erase' by default, it is possible to copy M0 and M4 images separately. Also e.g. IAR does not perform chip erase, so development in IAR IDE is possible provided the chip is first uploaded with M0 image.

theotherjimmy replied:
I appreciate the explanation. This is good to stay then!

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+
+    # update metadata
+    sig_bytes = ihex.tobinarray(start=0x90500002, size=4)
+    signature = (sig_bytes[0] << 24) | (sig_bytes[1]  << 16) | (sig_bytes[2] << 8) | sig_bytes[3]
```
Would struct.unpack work better here?

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+    for feature in image_features:
+        if not (feature in target_features):
+            return False
+    return True
```
This function could also be written as:
```python
return all(feature in target_features for feature in image_features)
```

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+        for image in m0_images:
+            features = image['features']
+            if check_matching_features(image['features'], target_features):
+                for hexf in resources.hex_files:
```
`resources.hex_files` is the names of all of the hex files in the build. What you want is the paths for every hex file in the build. This is available as:
```python
from tools.resources import FileType
resources.get_file_paths(FileType.HEX)
```
The current code will fail in the online compiler.

theotherjimmy commented on tools/targets/PSOC6.py:
```diff
@@ -0,0 +1,167 @@
+                break
+
+    if m0hexf:
+        toolchain.notify.info("M0 core image file found %s." % os.path.basename(m0hexf))
```
Please make this a `toolchain.notify.debug`.

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -625,7 +625,8 @@ def link_program(self, r, tmp_path, name):
                      if l.endswith(self.LIBRARY_EXT)]
-        dependencies = objects + libraries + [linker_script] + config_file
+        hex_files = r.get_file_paths(FileType.HEX)
+        dependencies = objects + libraries + [linker_script] + config_file + hex_files
```
This was probably quite helpful for development, but won't really change anything for a user. This is fine to stay in, as there's not really a downside that I can think of.

lrusinowicz replied:
I would defend this (and only this one), as in the dual-core application case, event with manual handling, e.g. running:
`$ mbed compile -m FUTURE_SEQUANA_M0 && cp BUILD/ FUTURE_SEQUANA/GCC_ARM/application.hex ./m0_images && mbed compile -m FUTURE_SEQUANA
`
this will prevent the necessity to manually delete previous build result. OK?

theotherjimmy replied:
I was not attacking this one, in fact I like it. It's fine to come in. Sorry that I put it in a request changes review.

theotherjimmy reviewed (changes requested):
I'm not convinced that the image merging present in this PR is something we should take. My concerns are:
 * Infrastructure is added to the exporters to support a probably-broken Makefile export.
 * This will break managed bootloader mode if you ever decide to turn that on.
 * I don't know how a user would add a new M0 firmware without modifying Mbed OS.

0xc0170 commented on drivers/Timer.cpp:
```diff
@@ -23,11 +23,13 @@ namespace mbed {
 
 Timer::Timer() : _running(), _start(), _time(), _ticker_data(get_us_ticker_data()), _lock_deepsleep(true)
 {
+    (void)ticker_read_us(_ticker_data); // Make sure h/w timer is initialized in non-critical context.
```
This fix should be at least a separate pull request. best if it's a separate pull request to have it reviewed separately from adding new target.

MarceloSalazar replied:
Indeed, this is something that affects multiple platforms, so the change shouldn't be introduced on this PR.
@lrusinowicz please remove this and send a separate PR if really needed. Thanks!

lrusinowicz replied:
I've created new pull request for this fix: #8502
Please make sure that it is merged before this one ( #8491 ).

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M0/device/TOOLCHAIN_IAR/cy8c6xx7_cm0plus.icf:
```diff
@@ -0,0 +1,218 @@
+/***************************************************************************//**
```
I'll add a comment here as I can't comment in the above .a  files.

`https://github.com/ARMmbed/mbed-os/tree/16ee057d462219ac9b7069fde1021c06946119df/targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M0/device` this folder should contain readme + license (PBL preferably) for binaries there. 

Readme should address: What binaries are there, what tools versions they were compiled with, and all these binaries are I assume released with the same license (there will be separate license file provided).

lrusinowicz replied:
Is it ok to just reference PBL on the ARM site in the readme file? Or should I include license file with PBL text copied in?

0xc0170 replied:
Just add a license as well please (License.txt). It's already present in other similar binaries in the tree.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M4/ipc_rpc.cpp:
```diff
@@ -0,0 +1,125 @@
+}
+
+
+#define RPC_GEN     RPC_GEN_INTERFACE_IDS_INIT
```
I am curious what magic is this with RPC_GEN and repeating including header files with the macro changed below

lrusinowicz replied:
I wanted to have automatic wrapper function generation (to Remote Procedure Call implementation) for a number of APIs to avoid cutting/pasting of the code. Technically, each wrapper function just copies its arguments into the buffer and calls a common communication function.
Yes, in C it's a kind of magic (I mean non-pretty). So I assume you would like to see more extensive comments on what's going on here, and I will add them to the source code.

0xc0170 replied:
That might help a reader

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M4/ipc_rpc.cpp:
```diff
@@ -0,0 +1,125 @@
+
+#undef RPC_GEN
+
+extern "C" {
```
why are these functions defined with C linkage here?

lrusinowicz replied:
I  have done it for consistency, because functions that are RPC-ed to M0 (from psoc6_utils.h) are low level ones, called from HAL implementation, so they are necessarily C API, I also wanted this mechanism to be most generic, so it meant lower common denominator, i.e. C. The third reason is that wrapper generation is a code shared between M0 and M4 cores, and for pre-compiled binaries for M0 core we are providing we have to use PSoC Creator IDE from Cypress and it is C-only.
However, you are right the the wrapper functions themselves are compiled with CPP compiler on Mbed, so only wrappers APIs have to be declared as C-linkage.

0xc0170 replied:
Understood.

> However, you are right the the wrapper functions themselves are compiled with CPP compiler on Mbed, so only wrappers APIs have to be declared as C-linkage.

Fix to apply to only wrappers API? I did not notice which ones are.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/analogout_api.c:
```diff
@@ -0,0 +1,185 @@
+ * @param obj The analogout object
+ * @return A floating-point value representing the current voltage on the pin,
+ *     measured as a percentage
+ */
```
Why are we having doxygen comments here (copy paste from header file, if we improve documentation in the analogout header file, this might be outdated and hard to maintain if each target implementation does it). 

This is valid for any HAL implementation ( I can see it in every _api HAL code file).

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/spi_api.c:
```diff
@@ -0,0 +1,648 @@
+ * @param[in] obj The SPI peripheral to check
+ * @return non-zero if the peripheral is currently transmitting
+ */
+int  spi_busy(spi_t *obj) {
```
Can you run astyle on the HAL implementation (`{` should be on the new line). Currently astyle check ignores target folder but will be enabled in the next release most likely.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/i2c_api.c:
```diff
@@ -0,0 +1,692 @@
+int  i2c_stop(i2c_t *obj_in)
+{
+    i2c_obj_t *obj = OBJ_P(obj_in);
+//    Cy_SCB_I2C_MasterSendStop(obj->base, obj->timeout, &obj->context);
```
dead code, please remove

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/i2c_api.c:
```diff
@@ -0,0 +1,692 @@
+/** Send START command
+ *
+ *  This function is a dummy operation on PSoC
+ */
```
In this case, having this is OK as it is different from i2c_start documentation and explains why this is empty (regarding my earlier comment about having copy-paste API docs in the implementation)

0xc0170 commented:
@yanesca @hanno-arm Please review trng implementation

0xc0170 commented:
Depends on #8502 (label preceding PR set)

yanesca commented:
@0xc0170 I'll have a look at it!

After a superficial look I already noticed that this contribution contains crypto accelerator headers and binaries as well. I also noticed that they are not integrated with Mbed TLS, therefore applications using Mbed TLS interface won't make use of them. I just would like to make sure that we all are aware of this and are happy to go forward without the integration.

@0xc0170 @lrusinowicz Can you both confirm?

lrusinowicz commented:
@theotherjimmy 
I'll start from the end, as the answer is a simplest one.
> I don't know how a user would add a new M0 firmware without modifying Mbed OS.

Hex files are detected (and added to the resource list) anywhere in the [source] tree. So if one wants to use it's own image for M0 it can place it in e.g the program root directory or any subdirectories of it, excluding BUILD.
I've also added functionality to merging (hook) functions to look for hex into the parallel target BUILD directory (if "enable_dual_core_build" switch is turned on). This assumed, that user application intended to utilize both cores would have target sub-directories (TARGET_MCU_PSOC6_M0 and TARGET_MCU_PSOC6_M4 or TARGET_FUTURE_SEQUANA_M0/TARGET_FUTURE_SEQUANA) holding source code unique to particular core.
The build process would then create two binary trees, BUILD/FUTURE_SEQUANA and BUILD/FUTURE_SEQUANA_M0 and will look for a hex file into the latter when merging images during FUTURE_SEQUANA build.

> This will break managed bootloader mode if you ever decide to turn that on.

Certainly, bootloader functionality sound great and most likely we would to have it enabled some time in the future. This, however, will require resolving some strategy issues, as the PSoC 6 is unique in Mbed with it's dual core hardware, so this won't happen automagically. Likely, bootloader will have to be implemented on M0 core, and right now M0 is not the main target of Mbed OS for this platform.


> Infrastructure is added to the exporters to support a probably-broken Makefile export.

I've once had changes that allow user to build complete project for both cores with the assumptions mentioned above, and it . This, however, was previously rejected/vetoed as too complicated/cumbersome, possibly jeopardizing ARM plans to cover multicore targets. 
Maybe in this situation I should resort to just minimal implementation, having user to manually select an image for M0. This strategy would probably result in removing most of the offending code. The downsides would be:

- User will have to manually select different M0 binary for BLE applications by means of mbed_app.json. Probably not and issue as one already has to add BLE feature using the same file.
- Users willing to create dual-core applications would have to copy over binary result of M0 compilation out of the BUILD directory and, possibly, rename it (for consistency).

Does it sound as a better plan?

theotherjimmy commented:
> So if one wants to use it's own image for M0 it can place it in e.g the program root directory or any subdirectories of it, excluding BUILD.

I disagree. The post-build script looks into the array: https://github.com/ARMmbed/mbed-os/blob/16ee057d462219ac9b7069fde1021c06946119df/targets/targets.json#L4550-L4561 and picks one of those images if it's present: https://github.com/ARMmbed/mbed-os/blob/16ee057d462219ac9b7069fde1021c06946119df/tools/targets/PSOC6.py#L144-L152 This makes the built-in images the only options as the post-build script raises a `ConfigException` if it can't find a built-in image,  and it's a crap shoot if you name your file the same as a built in image: https://github.com/ARMmbed/mbed-os/blob/16ee057d462219ac9b7069fde1021c06946119df/tools/targets/PSOC6.py#L154-L158 My understanding of the code is that you can't provide a new image to the post-build script without enabling this `sub-target-build-enabled` configuration parameter and building the image yourself https://github.com/ARMmbed/mbed-os/blob/16ee057d462219ac9b7069fde1021c06946119df/tools/targets/PSOC6.py#L120-L128 which does not work with custom build directories specified with `--build`.

A big issue I have with this post-build script is that it seems to be targeting a use case that is very similar to managed bootloader mode. Adding a second, target-specific merging functionality is a disadvantage for us in the long run as we would have an inconsistent story. Our story would be something like "use managed bootloader mode configuration parameters, unless your using a PSoC6, then use this other set of parameters that do nothing on every other target. Oh and further, your manager bootloader configuration will break the PSoC6". It's much easier for us to document and developers to use if the story is "use managed bootloader mode, it works on every target that can bootloader". My preference in this situation is to add a minimal target implementation and work or creating configuration to describe the nuance that your target adds to managed bootloader mode. I suspect that your target will not be alone in it's nuance, and that other targets will be able to take advantage of new features we add.

> as the PSoC 6 is unique in Mbed with it's dual core hardware, so this won't happen automagically. Likely, bootloader will have to be implemented on M0 core, and right now M0 is not the main target of Mbed OS for this platform.

The bootloader mode has no requirement that the image be built for the same target. The only requirements are that the image specified to `target.bootloader_img` starts the application that you're compiling now, and looks at the flash erase block after it's end for the vector table of the current application.

> Does it sound as a better plan?

Yes. We should have an offline discussion about what's special about your target and how we can add some general-purpose tools features that make it easier for developers to take advantage of these special features. For reference, the tools and post build scripts have to work with older versions of Mbed OS, meaning that any new features you add will have to be compatible with this post-build script.

------------------

Supposing that you can enable bootloader functionality now, we can alleviate the first downside you mention by:

 * Configuring the "just bootloder" M0 image as the bootloader in `targets.json`.
 * Adding an `mbed_lib.json` within a `FEATURE_BLE` directory that overrides this configuration with the BLE bootloader + stack
 * Allowing the user to specify their own `target.bootloader_img` in their `mbed_app.json`

This would at least remove the need to configure the bootloader, and retain the ability to use the bootloader of their choosing.

lrusinowicz commented:
> I disagree. The post-build script looks into the array: 

One would just need to overload array definitions by means of mbed_app.json. I haven't mention this, as it will always be the case.

The other thing is, that the code running on M0 is not a bootloader. Not only it's different functionally, but right now it doesn't even provide bootloading capabilities. Memory allocation schema is also different. And so on... So calling it a bootloader will just confuse developers.
Unfortunately, adding a bootloader capabilities to SEQUANA/PSOC6 target at this moment is not part of the plan.

Also, there are many targets having special post binary handling (at least all those having hook classes in tools/targets/\__init\__.py) or unusual parameters. I understand, that you are opposing creation of yet another special case, so I'm working on reducing it to the bare minimum. Hopefully, we can develop better (more consistent) way to handle this further down the road, maybe even fixing exporters as a side benefit. :)

lrusinowicz commented:
@yanesca 

> I also noticed that they are not integrated with Mbed TLS, therefore applications using Mbed TLS interface won't make use of them.

My understanding is that it's exactly the case. TLS is a different use case, on which Mbed TLS developers are working right now. I can remove TRNG implementation if you believe it will be troubling; we are not using it right now.

yanesca commented:
@lrusinowicz I don't think that it is troubling I just wanted to make sure that this is what you want.

I wanted to make it sure because TRNGs are handled differently from crypto accelerators in Mbed OS:
- TRNGs are integrated in HAL
- Crypto accelerators are integrated in Mbed TLS (which is not just a TLS stack, but a cryptographic library as well and most applications use it as a uniform interface to hardware accelerators instead of the drivers directly).

lrusinowicz committed:
Simplified M0/M4 binary merging functionality. Now, M0 binary image to be used has to be explicitely named in a json file (can be ovverriden in mbed_app.json).
Exporter hooks removed completely.
Cleanup and improvements to the comments, including removal of the redundant doxygen comments.
Code run through astyle. Additionally:
 - changes to drivers/Timer.cpp reverted
 - ipcpipe_transport.* files removed as they are not used for now,
 - fixed condition in stdio_init.cpp to perform serial initialization only when STDIO is enabled,
 - added missing resurce manager call in PWM initialization,
 - us_ticker initialization changed to use pre-reserved clock divider (to avoid resource manager call).
Changed reporting level from info to debug in PSOC6.py.
Added missing includes for function declarations in startup files.
Fixed (removed) garbadge text in psoc6_utils.c
Precompiled binaries updated for recent changes in psoc6_utils.c and moved to a separate folder; README and LICENSE files added.

theotherjimmy commented:
> Not only it's different functionally, but right now it doesn't even provide bootloading capabilities.

Managed bootloader mode is really just a way to:
 * Reserve space for an application than resides before the current application in ROM,
 * Merge the bootloader and application as part of a post build step,
 * In a target-independent way.

Even if you don't call the M0 code a bootloader, it seems like these steps would be useful for developers using your target. These steps should be especially useful as they understand flash erase blocks and will round up/down to the nearest one as needed. If we can generalize parts of your target's nuances, it may be possible for other targets to take advantage of these new features. For example, reserving RAM for the M0 could use the same mechanism as reserving RAM for a Nordic SoftDevice or a v8m secure partition.

> Memory allocation schema is also different.

Please elaborate. That would be useful information.

>So calling it a bootloader will just confuse developers.

Sure, I understand that unexpected terms can be confusing. What would you call it? a "SoftDevice" like Nordic? a "secure partition" like v8m?

We can alias the managed bootloader mode configuration so that it matches the terms you expect or better suite the use case.

> I understand, that you are opposing creation of yet another special case, so I'm working on reducing it to the bare minimum.

Reducing the special cases now will help get it in faster, as it will remove this discussion from the list of blockers. We certainly should work to maximize the support of your target, and I think that improving support can come as follow on PRs.

>Hopefully, we can develop better (more consistent) way to handle this further down the road, maybe even fixing exporters as a side benefit. :)

Agreed! :)  We should develop a more consistent way to handle this, and I'm willing to support extra exporter infrastructure if it comes with consistent support across all targets. I'll even collaborate on the tooling changes.

lrusinowicz commented:
I believe I've addressed all the issues with the latest set of commits.

@0xc0170 you can remove the PR dependency tag. I have changed our implementation of us_ticker and resource manager a bit to have all the resources for us_ticker pre-reserved. This way us_ticker_init() is safe to be called from critical section now and we no longer need Timer changes.

I'm attaching test results for the latest version.
[tests_sequana_arm_20181027.log](https://github.com/ARMmbed/mbed-os/files/2521931/tests_sequana_arm_20181027.log)
[tests_sequana_gcc_20181027.log](https://github.com/ARMmbed/mbed-os/files/2521932/tests_sequana_gcc_20181027.log)
[tests_sequana_iar_20181027.log](https://github.com/ARMmbed/mbed-os/files/2521933/tests_sequana_iar_20181027.log)

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/device/cymetadata.c:
```diff
@@ -0,0 +1,58 @@
+* This file is automatically generated by PSoC Creator.
+*
+********************************************************************************
+* Copyright (c) 2007-2017 Cypress Semiconductor.  All rights reserved.
```
I noticed all files have this copyright notice, but missing license file ? I could not locate any license file in folders. What is the license for these driver files ?

lrusinowicz replied:
I've added README and a license file (copied verbose from Cypress PDL). Is it ok now?

0xc0170 replied:
Thanks, Readme looks good to me. we will review the license

I noticed also some libs precompiled (ble or related), that does not have any license (the latest update covers hex/ folder ones. Please review

lrusinowicz replied:
These are also parts of the Cypress PDL. I've added more readme and license files to cover.

0xc0170 reviewed (dismissed):
Files libcy_crypto_client_mdk.a and similar also require license to be included as mentioned earlier - still cant find it here and what theyactually provide (simple readme would be helpful).

lrusinowicz committed:
Added missing README files for Cypress PDL library.
hex files subfolder moved up one level to avoid license confusion.
Removed non-TLS implementation of TRNG. Removed unused crypto libraries and headers.
Replaced Cypress copyright licence per agreement.
Removed unsed eeprom emulation middleware files.
Renamed assembler files from *.s to *.S
Removed "device_name" from targets.json definitions as it is not supoprted yet.

hanno-becker commented:
@0xc0170 Could you remove the review request from me? @yanesca has taken over the crypto review work again.

theotherjimmy reviewed (approved):
This is much simpler, and directly compatible with managed bootloader mode when `target.sub_target` is not set. While I'm not convinced that `target.sub_target` is the correct name, it's optional and not worth blocking on.

theotherjimmy commented on targets/targets.json:
```diff
@@ -4489,5 +4489,83 @@
+    },
+    "MCU_PSOC6_M4": {
+        "inherits": ["MCU_PSOC6"],
+        "macros": ["CY_CORE_ID=0", "MCU_PSOC6_M4"]
```
Something seems amiss: The M4 and M0 have the same "CORE_ID".

MarceloSalazar replied:
Now fixed.

lrusinowicz replied:
Fortunately CY_CORE_ID was like a leftover from the older version, now it's not used anymore so I just deleted it.

MarceloSalazar commented:
@0xc0170 can you remove the TLS label, please?

adbridge commented:
@yanesca Could you please take another look at this and check they have cleanly removed the TRNG stuff? Thanks

yanesca reviewed (approved):
@adbridge , @MarceloSalazar I can confirm that all the crypto code has been removed.

(Some code remained for setting up the crypto accelerator interrupt handlers, but as far as I can tell  this shouldn't cause any problems.)

cmonr commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8C63XX/TARGET_MCU_PSOC6_M0/system_psoc63_cm0plus.c:
```diff
@@ -0,0 +1,636 @@
+* \copyright
+* Copyright 2016-2018, Cypress Semiconductor Corporation.  All rights reserved.
+* Copyright 2017-2018, Future Electronics
+ */* SPDX-License-Identifier: Apache-2.0
```
```suggestion
 * SPDX-License-Identifier: Apache-2.0
```

I think the '*/' might cause some issues here.

adbridge replied:
It most definitely will. Line 1 has /* and matching * / and then another /* which originally was being matched by the */ on line 11 but then there is an extra */ on line 12.

lrusinowicz replied:
I have fixed this typo with the latest commit.

cmonr reviewed (commented):
Quick note on the latest commit. Looks like one of the changes is causing github parsing to be thrown off, which might indicate build errors once we start CI going.

adbridge commented:
The licences all now seem to be correct.

lrusinowicz commented:
@adbridge 
Could you please comment on why the travis-ci/docs tests have failed (what's wrong with assembler file names)? Thanks.

cmonr commented:
@lrusinowicz There's a requirement that .a and .ar files start with the `lib` prefix. I'm personalyl not sure why tho.

cmonr commented:
Still investigating the python failure, but @lrusinowicz it appears that at some point, a merge commit was made against master, which is throwing off the commit history.

For example, this commit should only exist in master, but it's showing up here as well:
> @cmonr
Merge pull request #8519 from cmonr/rollup

cmonr commented:
@lrusinowicz One other thing. Even if we get all of the tests green, this PR will not be able to come in because it's not properly following our process.

At some point, a merge against master was pushed into this PR, which wrecks havoc with the PR history. As you can now see, the commit history for this PR now contains commit and authors that aren't even aware of this PR. As I mentioned in my previous comment, there's a commit in this PR that I'm absolutely sure should not exist in this PR.

I suspect it was this commit, but I'm not 100% sure.
> Merge branch 'new-target-future_sequana' of github.com:lrusinowicz/mb…
…ed-os into new-target-future_sequana


What will need to happen, is that the merge commit that is causing the incorrect history to be shown, will need to be removed, _and then_ in order to update the PR, it will need to be rebased.

### Removing the offending merge commit
Note: This assumes that the commit I'm singling out is the correct commit.
The following commands should remove the offending commit.
```
# Backup the curernt repo, because backing up is _always_ a good idea
cp -r mbed-os mbed-os.bak

cd mbed-os
git checkout new-target-future_sequana

# Remove the last two commits from the history.
#  Two, since the assembly file rename was added. That commit will need to be redone, but one step at a time.
git reset --hard HEAD~2
```

### Updating the PR
```
git remote add upstream https://github.com/ARMmbed/mbed-os
git fetch upstream master
git rebase upstream/master

# At this point, git will replay the commits in this branch on top of the current ARMmbed/mbed-os master HEAD. Conflicts will likely arise. Follow the git instructions.
```

### Wrapping up
Once the rebase is complete, the final step is a `git push --force`.
The `--force` parameter is needed since we've rewritten the history, and the local branch is now different than the remote branch, in a way that git does not konw how to solve. The `--force` means that we're sure we want the local changes to replace the branch changes.

cmonr commented:
@lrusinowicz Feel free to ask any questions here. I'm still poking around to figure out why the last travis python failure occurred, but I think I have an idea as to why.

lrusinowicz commented:
@cmonr
I hope I cleaned thing up regarding offending merge, thank you very much.
I've found a reason for a one of the travis tests failure, assembler files named with a small 's' suffix instead of capital 'S' and I have fixed that.
Unfortunately, I'm afraid I cannot fix the other failure. It seems to have something to do with flash writing/binary blobs/cmsis packs?
I know what binary blobs are, but I don't think we need one to write flash memory on PSoC 6, the greentea flash tests are passing OK on our target.

cmonr commented:
@lrusinowicz Thank you! The commit history appears much better now. Part of my concern was that the number of changes seems so high, but then again, it's not every day we get a new vendor.

Thanks for catching and fixing that .s/.S shenanegans as well.

The lone failure appears to be something in regards to the target naming, but I'm still not sure what it could be.

cmonr commented:
@lrusinowicz Question.
Is there a CMSIS pack associated with this device? I'm not sure the `"device_name": "CY8C6347BZI-BLD53M0+"` line is needed if that isn't the case, and think the test is complaining because it can't find a pack for it.

lrusinowicz commented:
@cmonr 
I could probably squash some of the recent commits while rebasing, but I was not sure (and was a bit short on time :smiley: ). Do you think I should do it yet?

cmonr commented:
@lrusinowicz That's fine if you want to go for it :)

cmonr commented on targets/targets.json:
```diff
@@ -4533,5 +4533,82 @@
+    "FUTURE_SEQUANA_M0": {
+        "inherits": ["MCU_PSOC6_M0"],
+        "supported_form_factors": ["ARDUINO"],
+        "device_name": "CY8C6347BZI-BLD53M0+",
```
```suggestion

```

cmonr replied:
Removing these two line will allow the Travis CI tests to pass. 

I created a clone of this PR in my personal fork, which is also setup in Travis, and confirmed it.
Before: https://travis-ci.org/cmonr/mbed-os/jobs/449536076
After: https://travis-ci.org/cmonr/mbed-os/jobs/449539603

cmonr commented on targets/targets.json:
```diff
@@ -4533,5 +4533,82 @@
+        "inherits": ["MCU_PSOC6_M4"],
+        "sub_target": "FUTURE_SEQUANA_M0",
+        "supported_form_factors": ["ARDUINO"],
+        "device_name": "CY8C6347BZI-BLD53M4",
```
```suggestion

```

cmonr reviewed (commented):
Change needed before it can pass Travis

cmonr commented:
Thanks @bridadan for the hint!

lrusinowicz commented:
@cmonr @bridadan 
Thank you very much for helping to find and fix issue with travis tests!
I've also done some squashing - should be more acceptable now I hope. :smiley:

bridadan commented:
Not a problem @lrusinowicz! Just to confirm what @cmonr, does this target have a CMSIS pack?

cmonr commented:
@bridadan I didn't find one in the Keil software downloads section, so I don't think so.

lrusinowicz commented:
@bridadan 
I believe no, Cypress PSoC 6 chips do not have yet official CMSIS pack.

cmonr commented:
So far, it looks like all of our automatic CI checks are going well, so I'm going to start the full CI as well to get it going in parallel.

Fingers crossed!
/morph build

cmonr commented:
@lrusinowicz A heads up, that if the Cypress devices are only supported in IAR 8, the IAR export of the devices will probably fail. We should know in around 2 hours.

lrusinowicz commented:
Just for the reference I'm attaching latest greentea tests results. 
[tests_sequana_arm_20181101.log](https://github.com/ARMmbed/mbed-os/files/2540230/tests_sequana_arm_20181101.log)
[tests_sequana_gcc_20181101.log](https://github.com/ARMmbed/mbed-os/files/2540231/tests_sequana_gcc_20181101.log)
[tests_sequana_gcc_20181101_2.log](https://github.com/ARMmbed/mbed-os/files/2540232/tests_sequana_gcc_20181101_2.log)
[tests_sequana_iar_20181101.log](https://github.com/ARMmbed/mbed-os/files/2540233/tests_sequana_iar_20181101.log)
GCC results are split into two because I've forgot to connect the target board. :smiley:

cmonr commented:
@lrusinowicz We're looking into why the builds auto-aborted (duration time limit), but the build logs also have build failures. Could you take a look at those please?

lrusinowicz commented:
@cmonr 
I have trouble interpreting the results.
One thing I see for sure is that "intermediate" targets MCU_PSOC6, MCU_PSOC6_M0 and MCU_PSOC6_M4 are not supposed to be build directly. This might have been missed out in a review.
But I'm not sure why FUTURE_SEQUANA and FUTURE_SEQUANA_M0 are reported as failures?

lrusinowicz committed:
Removed "intermediate" PSOC6 targets from buildable target list.

cmonr commented:
That'll do it!
/morph build

cmonr commented:
@lrusinowicz Looks like there are about to be more errors.

> TypeError: write_hex_file() got an unexpected keyword argument 'byte_count'

http://mbed-os.s3-eu-west-1.amazonaws.com/builds/8491/FAIL/FUTURE_SEQUANA_M0/ARM/6203d04630f0d77f4725ff279c39fb037b6c0a9a_build_log_FUTURE_SEQUANA_M0_ARM.txt

http://mbed-os.s3-eu-west-1.amazonaws.com/builds/8491/FAIL/FUTURE_SEQUANA/GCC_ARM/6203d04630f0d77f4725ff279c39fb037b6c0a9a_build_log_FUTURE_SEQUANA_GCC_ARM.txt

cmonr commented:
The errors appear to be centered around the PSOC-specific build scripts:

> File "/builds/ws/mbed-os-build-matrix/3529/FUTURE_SEQUANA_M0_ARM/sources/mbed-os/tools/targets/PSOC6.py", line 94, in complete_func
    ihex.write_hex_file(dest if dest else hexf0, write_start_addr=False, byte_count=64)

MarceloSalazar commented:
> TypeError: write_hex_file() got an unexpected keyword argument 'byte_count'

I confirm I was seeing the same issue and believe can be fixed by upgrading intelhex to 2.2.1 in the CI.
See more: https://github.com/bialix/intelhex/blob/master/NEWS.rst

cmonr commented:
K. It appears that action is needed on our side to get this through. 

Alternatively, if it's possible to depend on an older IntelHex module, and update this PR, we might be able to progress that way as well.

cmonr commented:
Ticket created: IOTTESTINF-4642

adbridge commented:
/morph build

adbridge commented:
Something strange going on in CI with this, trying again.

/morph build

cmonr commented:
@lrusinowicz There's a lot of noise in the results because it looks like we had some exporter node issues, but please take a look at the following logs:
- [make_iar](http://mbed-os.s3-eu-west-1.amazonaws.com/builds/exporter/8491/FAIL/FUTURE_SEQUANA/make_iar/6203d04630f0d77f4725ff279c39fb037b6c0a9a_exporter_build_log_FUTURE_SEQUANA_make_iar.txt)
- [make_gcc](http://mbed-os.s3-eu-west-1.amazonaws.com/builds/exporter/8491/FAIL/FUTURE_SEQUANA/make_gcc_arm/6203d04630f0d77f4725ff279c39fb037b6c0a9a_exporter_build_log_FUTURE_SEQUANA_make_gcc_arm.txt)
- [make_armc5](http://mbed-os.s3-eu-west-1.amazonaws.com/builds/exporter/8491/FAIL/FUTURE_SEQUANA/make_armc5/6203d04630f0d77f4725ff279c39fb037b6c0a9a_exporter_build_log_FUTURE_SEQUANA_make_armc5.txt)

All of them failed with the message `file contains no data`. Suspecting something is wrong with how the new target is interacting with the makefiles that make up these compilers.

cmonr commented:
@lrusinowicz I was reminded that the command that shows up in the log involves some steps that are prepared by the CI...

If the quickest thing for you is to simply disable the file exporting, which can be confirmed by running the `mbed export -S` command, and verifying that FUTURE_SEQUANA and FUTURE_SEQUANA_M0 don't have any `make_x` checked.

lrusinowicz commented:
It looks like indeed, the srec_cat has problem interpreting hex files. But this result I only get when I run srec_cat without properly specifying file format with -intel option. When I run it with a command line optione as specified in a makefile it works properly.
Which version of srec_cat utility do you have installed on the servers?

cmonr commented:
@lrusinowicz What python module package is this a part of? I've never head of the utility before.

cmonr commented:
@lrusinowicz Nevermind. Found the utility you were refering to.
Would you mind providing the version you're using to develop with? I can ask someone who has access to see what version we have.

lrusinowicz committed:
Disabled makefile exporters for PSoC 6 based targets.

cmonr commented:
@lrusinowicz Fwiw, I compiled a recent version for my work VM (Arch Linux), and found that make (after running mbed export -i make_gcc_arm -m FUTURE_SEQUANA) still failed.

cmonr commented:
@lrusinowicz When you're ready, I can restart CI with the updated commit.

Already locally verified that the three failed exporters should be disabled.

lrusinowicz commented:
@cmonr 
Yes, I was also able to reproduce the issue locally. I've already submitted a patch disabling makefile exporters.

lrusinowicz commented:
@cmonr 
I've verified that the only exporters reported as enabled for SEQUANA targets are codeblocks, e2studio and gnuarmeclipse.
Codeblocks exporter works, as I'm using it daily.
Running gnuarmeclipse doesn't report errors, but the generated project doesn't seem to be correct - eclipse doesn't complain, but the project seems to be empty.
I have no idea about the e2studio.

cmonr commented:
/morph build

cmonr commented:
/morph mbed2-build

cmonr commented:
I was starting to get concerned since the exporter CI tests generally take no more than two hours when started, but was relieved to see that the job is _just_ being starved of Windows Nodes due to the autotriggering of other export jobs.

The jobs that have completed thus far are all green. Unable to provide an eta atm, but holding 5.10.3 PR generation for this PR.
